src/Tools/Pagination/Paginator.php

```diff
@@ -51,6 +63,29 @@ public function __construct(
         $this->query = $query;
     }
 
+    /**
+     * Create an instance of Paginator with auto-detection of whether the provided
+     * query is suitable for simple (and fast) pagination queries, or whether a complex
+     * set of pagination queries has to be used.
+     */
+    public static function newWithAutoDetection(QueryBuilder $queryBuilder): self
+    {
+        /** @var array<string, Join[]> $joinsPerRootAlias */
+        $joinsPerRootAlias = $queryBuilder->getDQLPart('join');
```

Comment:
would it be possible to also support `Query` by analyzing the DQL AST instead of the QueryBuilder ? this would make this method more useful.

Reply:
1. I would prefer to support `Query` _in addition to_ `QueryBuilder`, not _instead of_. This is because constructing an AST from a Query's DQL is not instantaneous, while checking for presence of DQL parts in the QueryBuilder is.
2. My main use case for the Paginator (and I would assume that of a lot of other real-world usages too), is to implement CRUD listings that support filtering/conditions. The conditions are only added when selected by the user in the website gui. Since the conditions are added in a "conditional manner", the DQL query that retrieves the records for the CRUD listing is often built incrementally via a QueryBuilder and a bunch of if-statements (in one form or another -- I actually use a Symfony Bundle for the "bunch of if-statement" part). My point being: supporting the instances of Query is not my personal main priority. I could look into doing it, if I have the time after getting the "initial approval" for this PR (and I still have the unit tests to write).
3. Supporting `Query|QueryBuilder` is definitely an option.

Are you ok with this new function's supporting only instances of QueryBuilder if I decide that I ran out of time for this PR?

Reply:
@d-ph Supporting `QueryBuilder` in addition to `Query` would then work the same than what is done in the constructor: calling `getQuery` on it and then dealing with the Query.

I'm not asking for the public API to drop support for QueryBuilder.

Reply:
Supporting queries is important to support cases using `Query::setFetchMode` to trigger eager loading of some `toOne` relations (which would load them in batch in a single query instead of doing N lazy-loading queries).
In my experience, this can result in better performance in some cases compared to fetch-joining the relation in the paginated query (assuming the paginated query does not already join it for filtering purposes of course):
- if the toOne relation is nullable and is often `null`
- if you have multiple `toOne` relations pointing to the same entity (the batch loading is per target entity, not per relation, so they would be grouped)
- if the target entity has a non-negligeable probably to refer to an instance that is already initialized in the identity map (if the current page only contains values using `null` or an initialized entity, the eager loading won't need to perform a second query at all).

The QueryBuilder cannot configuring the fetch mode of relations. So even when using the QueryBuilder (which I often do), I often have cases where I pass a `Query` to the Paginator instead of passing the QueryBuilder.

Reply:
> Supporting `QueryBuilder` in addition to `Query` would then work the same than what is done in the constructor: calling `getQuery` on it and then dealing with the Query.

What about the comment I raised in point 1. above about the cost of constructing a Query's AST?

My intention was to at some point add the support for Query, but in a form where there would be separate routines for inspecting the provided instance of Query or QueryBuilder, so that the AST construction is avoided when a QueryBuilder is provided. Something along the following lines:
```
$sqlJoinsPresent = false;

if ($queryOrQueryBuilder instance of Query) {
  $ast = $queryOrQueryBuilder->getAST();
  // inspect the AST
  $sqlJoinsPresent = true;
} else {
  // do not construct any ASTs
  // use QueryBuilder::getPart(partName) to inspect the query 
  $sqlJoinsPresent = true;
}
```
Are you saying that you'd like to disregard the fact that constructing an AST is not a cheap operation (and which depends on the complexity of the query)?

---

> The QueryBuilder cannot configure the fetch mode of relations.

👍

Reply:
I could accept a fast path introspecting the QueryBuilder instead of parsing the Query as an optimization if it has indeed an impact. However, it would have to support the same capabilities. See my other comment related to the detection of joining **collections** rather than any joins. Maybe the QueryBuilder introspection would not be viable for that.

However, my personal opinion would be that we should not merge this without support for Query as it would prevent downstream projects from migrating to it (for instance, the Pagerfanta DoctrineOrmAdapter uses this Paginator and currently supports `Query|QueryBuilder`).
Note that this is only my own opinion and the ORM maintainers might take a different decision (I'm maintainer on DoctrineBundle but not on the ORM itself, even though I regularly help with reviews here as well)

Reply:
@stof when there are multiple to-one associations and the eager loading fetch mode is used, does this cause multiple result rows for a single entity and this is something where we need to be careful with regard to the goal of this PR?

Reply:
to-one associations don't create multiple rows in the SQL results. Only `toMany` relations do that.

Reply:
Right, even when there are multiple of them 🤦🏻‍♂️